polymer/src/sample-category-data.html

```diff
@@ -0,0 +1,203 @@
+
+        categories: {
+          type: Array,
+          notify: true
+        },
+
+        category: {
+          type: Object,
+          //computed: '_computeCategory(categoryName)',
+          notify: true
+        },
+
+        item: {
+          type: Object,
+          computed: '_computeItem(category.items, itemName)',
+          notify: true
+        },
+
+        failure: {
+          type: Boolean,
+          notify: true,
+          readOnly: true
+        }
+
+      }}
+
+      static get observers() { return [
+        '_categoriesLoaded(categories)', '_categoryNameSet(categoryName)'
+      ]}
+
+      _categoriesLoaded(categories) {
+        this._computeCategory(this.categoryName)
+      }
+
+      _categoryNameSet(categoryName) {
+        this._fetchCategories(3);
+      }
+
+
+      _getCategoryObject(categoryName) {
+        if (this.categories === undefined) {
+          return null;
+        }
+        for (let i = 0, c; c = this.categories[i]; ++i) {
+          if (c.id === categoryName) {
+            this.category = c;
+            return c;
+          } else if (categoryName === "index.html") {
+            return null
+          }
+        }
+      }
+
+      _computeCategory(categoryName) {
+
+        // Fetch the items of the category. Note that the fetch is asynchronous,
+        // which means `category.items` may not be set initially (but that path
+        // will be notified when the fetch completes).
+        let categoryObj = this._getCategoryObject(categoryName);
+        this._fetchItems(categoryObj, 1);
+        return categoryObj;
+      }
+
+      _computeItem(items, itemName) {
+        if (!items || !itemName) {
+          return;
+        }
+        for (let i = 0, item; item = items[i]; ++i) {
+          if (item.url.endsWith(itemName.concat('/'))) {
+            return item;
+          }
+        }
+      }
+
+      handleResponse(data) {
+        var cats = data.detail.response[0]["categories"]
+        var catNames =  cats.map(function(cat) {
+          var obj = {
+            name:  cat["name"],
+            id: cat["id"]
+          };
+          return obj;
+        });
+        this.categories = catNames;
+      }
+
+      _fetchCategories(attempts) {
+        this._setFailure(false);
+        if (this.categories) {
+          return;
+        }
+        this._getResource({
+          url: 'https://ampbyexample.com/sitemap.json',
+          onLoad(e) {
+            var cats = JSON.parse(e.target.responseText)[0]["categories"]
+            var catNames =  cats.map(function(cat) {
+              var obj = {
+                name:  cat["name"],
+                id: cat["id"]
+              };
+              return obj;
+            });
+            this.set('categories', catNames);
+            this.categories = catNames;
+          },
+          onError(e) {
+            this._setFailure(true);
+          }
+        }, attempts);
+      }
+
+      _fetchItems(category, attempts) {
```

Comment:
There is quite some duplication with the previous method. Maybe extract the common logic or reuse the `e.target.responseText` instead of sending multiple requests to the server.

Reply:
I extracted a variable for the sitemap URL and a method for getting the categories. WDYT?

Reply:
I think instead, the category-data should fetch the json once and then you use the result whenever you need it. E.g. [`_getResource`](https://github.com/ampproject/amp-by-example/pull/930/files#diff-262d6dba3394cd2624cde3516ac168b7R166) is called only once inside `category-data`. Then when you need to use the data, you use the property as bound by `category-data` instead.

Reply:
Ok, I created a property where I am storing the response and then using it when fetching items. PTAL.

Reply:
@TimvdLippe friendly ping :-)

Reply:
Attempts is never used